runtime/src/iree/hal/utils/libmpi.h

```diff
@@ -13,6 +13,40 @@
 extern "C" {
 #endif  // __cplusplus
 
+//===----------------------------------------------------------------------===//
+// Data types and enumerations
+//===----------------------------------------------------------------------===//
+
+// OpenMPI uses pointers for common data types instead of `int` like most other
+// implementations. This is unfortunate as when dynamically importing we can't
+// easily switch between the two calling conventions. Today we assume OpenMPI
+// everywhere but Windows where we use MSMPI which follows the normal spec.
+#if !defined(IREE_PLATFORM_WINDOWS)
+#define IREE_MPI_TYPES_ARE_POINTERS 1
+#endif  // !IREE_PLATFORM_WINDOWS
+
+#if IREE_MPI_TYPES_ARE_POINTERS
+
+typedef void* IREE_MPI_Datatype;
+#define IREE_MPI_BYTE(syms) (IREE_MPI_Datatype)((syms)->ompi_mpi_byte)
+
+typedef void* IREE_MPI_Comm;
+#define IREE_MPI_COMM_WORLD(syms) (IREE_MPI_Comm)((syms)->ompi_mpi_comm_world)
+
+#else
+
+typedef int IREE_MPI_Datatype;
+#define IREE_MPI_BYTE(syms) ((IREE_MPI_Datatype)0x4C00010D)
```

Comment:
I have not seen this value defined in the MPI 4.0 spec, so I'm afraid that this is  OpenMPI specific so there is not much benefit on defining it as such.

Reply:
This would be MPICH - if there's no consistency of these constants across implementations then we'll have other issues - we'll need to look into that. For now this at least allows MSMPI (MPICH-derived) to work. I'll add a comment.